messages/off-chain-registry/message_manager.go

```diff
@@ -0,0 +1,198 @@
+	data, err := json.Marshal(messageProtocolConfig.Settings)
+	if err != nil {
+		logger.Error("Failed to marshal off-chain registry config")
+		return nil, err
+	}
+	var messageConfig Config
+	if err := json.Unmarshal(data, &messageConfig); err != nil {
+		logger.Error("Failed to unmarshal off-chain registry config")
+		return nil, err
+	}
+
+	if err := messageConfig.Validate(); err != nil {
+		logger.Error(
+			"Invalid off-chain registry config.",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	return &messageManager{
+		logger:             logger,
+		destinationClients: destinationClients,
+		registryAddress:    common.HexToAddress(messageConfig.TeleporterRegistryAddress),
+	}, nil
+}
+
+// ShouldSendMessage returns false if the version/protocol address tuple is already registered in the TeleporterRegistry contract.
+func (m *messageManager) ShouldSendMessage(unsignedMessage *warp.UnsignedMessage, destinationBlockchainID ids.ID) (bool, error) {
+	addressedPayload, err := warpPayload.ParseAddressedCall(unsignedMessage.Payload)
+	if err != nil {
+		m.logger.Error(
+			"Failed parsing addressed payload",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	entry, destination, err := teleporterregistry.UnpackTeleporterRegistryWarpPayload(addressedPayload.Payload)
+	if err != nil {
+		m.logger.Error(
+			"Failed unpacking teleporter registry warp payload",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	if destination != m.registryAddress {
+		m.logger.Info(
+			"Message is not intended for the configured registry",
+			zap.String("destination", destination.String()),
+			zap.String("configuredRegistry", m.registryAddress.String()),
+		)
+		return false, nil
+	}
+
+	// Get the correct destination client from the global map
+	destinationClient, ok := m.destinationClients[destinationBlockchainID]
+	if !ok {
+		return false, fmt.Errorf("relayer not configured to deliver to destination. destinationBlockchainID=%s", destinationBlockchainID.String())
+	}
+	client, ok := destinationClient.Client().(ethclient.Client)
+	if !ok {
+		panic(fmt.Sprintf("Destination client for chain %s is not an Ethereum client", destinationClient.DestinationBlockchainID().String()))
+	}
+
+	// Check if the version/protocol address tuple is already registered in the TeleporterRegistry contract.
+	registry, err := teleporterregistry.NewTeleporterRegistryCaller(m.registryAddress, client)
+	if err != nil {
+		m.logger.Error(
+			"Failed to create TeleporterRegistry caller",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	address, err := registry.GetAddressFromVersion(&bind.CallOpts{}, entry.Version)
+	if err != nil {
+		// Check if execution reverted due to the version not being found
+		if strings.Contains(err.Error(), revertVersionNotFoundString) {
+			return true, nil
+		}
+		m.logger.Error(
+			"Failed to get address from version",
+			zap.Error(err),
+		)
+		return false, err
+	}
+	if address == entry.ProtocolAddress {
```

Comment:
If there is any address registered for this versions already, should we not send the message?

Reply:
Yup, that's indeed the case. We only send the message if the version is not found. In that case, we don't care if the address is already registered, since the same address can map to multiple versions. We can therefor eliminate the call to GetVersionFromAddress altogether.

Reply:
Yup, that's indeed the case. We only send the message if the version is not found. In that case, we don't care if the address is already registered, since the same address can map to multiple versions. We can therefor eliminate the call to GetVersionFromAddress altogether.